src/bin/customer/pay.rs

```diff
@@ -22,91 +23,137 @@ use super::{connect, database, Command};
-        let chan = zkabacus_pay(
-            rng,
-            database.as_ref(),
-            &self.label,
-            session_key,
-            chan,
-            payment_amount,
+        // Timeout is set to 10 messages, which includes all sent & received messages and aborts
+        let chan = tokio::time::timeout(
+            10 * config.message_timeout(),
+            zkabacus_pay(
+                rng,
+                database.as_ref(),
+                &self.label,
+                session_key,
+                chan,
+                payment_amount,
+            ),
         )
         .await
+        .context("Payment timed out while updating channel status")?
         .context("Failed to complete pay protocol")?;
 
-        // Receive the response note (i.e. the fulfillment of the service)
-        let (response_note, chan) = chan
-            .recv()
+        tokio::time::timeout(config.approval_timeout(), receive_service(chan))
             .await
-            .context("Failed to receive response note")?;
+            .context("Payment timed out when receiving service")??;
 
-        // Close the communication channel: we are done communicating with the merchant
-        chan.close();
+        Ok(())
+    }
+}
 
-        // Print the response note on standard out
-        if let Some(response_note) = response_note {
-            eprintln!(
-                "Payment succeeded with response from merchant: \"{}\"",
-                response_note
-            );
-        } else {
-            eprintln!("Payment succeeded with no concluding response from merchant");
-        }
+/// Parse the payment amount specified on the command line.
+fn parse_payment_amount(amount: Amount) -> Result<PaymentAmount, anyhow::Error> {
+    // Convert the payment amount appropriately
+    let minor_units: i64 = amount.try_into_minor_units().ok_or_else(|| {
+        anyhow::anyhow!("Payment amount invalid for currency or out of range for channel")
+    })?;
 
-        Ok(())
+    (if minor_units < 0 {
+        PaymentAmount::pay_customer
+    } else {
+        PaymentAmount::pay_merchant
+    })(minor_units.abs() as u64)
+    .context("Payment amount out of range")
+}
+
+/// Set up the communication channel with the merchant.
+async fn open_session(
+    database: &dyn QueryCustomer,
+    config: &Config,
+    channel_name: &ChannelName,
+) -> Result<(SessionKey, Chan<pay::Pay>), anyhow::Error> {
+    // Look up the address and current local customer state for this merchant in the database
+    let address = database
+        .channel_address(channel_name)
+        .await
+        .context("Failed to look up channel address in local database")?;
+
+    // Connect and select the Pay session
+    let (session_key, chan) = connect(config, &address).await?;
+    let chan = chan
+        .choose::<2>()
+        .await
+        .context("Failed selecting pay session with merchant")?;
+
+    Ok((session_key, chan))
+}
```

Comment:
Why extract this into its own function if it's only called once?

Reply:
Mostly to keep the main `run` function shorter and more readable. I prefer to keep that section as a series of clearly-named, high-level function calls, and this stuff felt a bit low level for it.